File: lca/src/main/java/com/dreampany/lca/data/model/Coin.java
package com.dreampany.lca.data.model;

import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;
import android.os.Parcel;
import androidx.annotation.NonNull;
import com.dreampany.frame.data.model.Base;
import com.google.common.base.Objects;
import com.google.common.collect.Maps;
import com.google.firebase.firestore.Exclude;
import com.google.firebase.firestore.IgnoreExtraProperties;

import org.apache.commons.lang3.builder.ReflectionToStringBuilder;

import java.io.Serializable;
import java.util.*;

/**
 * Created by Hawladar Roman on 29/5/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
@Entity(indices = {@Index(value = {"id"}, unique = true)},
        primaryKeys = {"id"})
@IgnoreExtraProperties
public class Coin extends Base {

    private long coinId;
    private String name;
    private String symbol;
    private String slug;
    private int rank;
    private int marketPairs;
    private double circulatingSupply;
    private double totalSupply;
    private double maxSupply;
    private long lastUpdated;
    private long dateAdded;
    private List<String> tags;
    @Ignore
    @Exclude
    private Map<Currency, Quote> quotes;

    @Ignore
    public Coin() {

    }

    public Coin(long id) {
        this.id = id;
    }

    @Ignore
    private Coin(Parcel in) {
        super(in);
        coinId = in.readLong();
        name = in.readString();
        symbol = in.readString();
        slug = in.readString();
        rank = in.readInt();
        marketPairs = in.readInt();
        circulatingSupply = in.readDouble();
        totalSupply = in.readDouble();
        maxSupply = in.readDouble();
        lastUpdated = in.readLong();
        dateAdded = in.readLong();
        tags = in.createStringArrayList();
        quotes = (Map<Currency, Quote>) in.readSerializable();
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeLong(coinId);
        dest.writeString(name);
        dest.writeString(symbol);
        dest.writeString(slug);
        dest.writeInt(rank);
        dest.writeInt(marketPairs);
        dest.writeDouble(circulatingSupply);
        dest.writeDouble(totalSupply);
        dest.writeDouble(maxSupply);
        dest.writeLong(lastUpdated);
        dest.writeLong(dateAdded);
        dest.writeStringList(tags);
        dest.writeSerializable((Serializable) quotes);
    }

    public static final Creator<Coin> CREATOR = new Creator<Coin>() {
        @Override
        public Coin createFromParcel(Parcel in) {
            return new Coin(in);
        }

        @Override
        public Coin[] newArray(int size) {
            return new Coin[size];
        }
    };

    @Override
    public int hashCode() {
        return Objects.hashCode(symbol);
    }

    @NonNull
    @Override
    public String toString() {
        return ReflectionToStringBuilder.toString(this);
    }

    public void setCoinId(long coinId) {
        this.coinId = coinId;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setSymbol(String symbol) {
        this.symbol = symbol;
    }

    public void setSlug(String slug) {
        this.slug = slug;
    }

    public void setRank(int rank) {
        this.rank = rank;
    }

    public void setMarketPairs(int marketPairs) {
        this.marketPairs = marketPairs;
    }

    public void setCirculatingSupply(double circulatingSupply) {
        this.circulatingSupply = circulatingSupply;
    }

    public void setTotalSupply(double totalSupply) {
        this.totalSupply = totalSupply;
    }

    public void setMaxSupply(double maxSupply) {
        this.maxSupply = maxSupply;
    }

    public void setLastUpdated(long lastUpdated) {
        this.lastUpdated = lastUpdated;
    }

    public void setDateAdded(long dateAdded) {
        this.dateAdded = dateAdded;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }

    public void setQuotes(Map<Currency, Quote> quotes) {
        this.quotes = quotes;
    }

    public void addQuote(Quote quote) {
        if (quotes == null) {
            quotes = Maps.newHashMap();
        }
        quotes.put(quote.getCurrency(), quote);
    }

    public long getCoinId() {
        return coinId;
    }

    public String getName() {
        return name;
    }

    public String getSymbol() {
        return symbol;
    }

    public String getSlug() {
        return slug;
    }

    public int getRank() {
        return rank;
    }

    public int getMarketPairs() {
        return marketPairs;
    }

    public double getCirculatingSupply() {
        return circulatingSupply;
    }

    public double getTotalSupply() {
        return totalSupply;
    }

    public double getMaxSupply() {
        return maxSupply;
    }

    public long getLastUpdated() {
        return lastUpdated;
    }

    public long getDateAdded() {
        return dateAdded;
    }

    public Date getLastUpdatedDate() {
        return new Date(getLastUpdated());
    }

    public List<String> getTags() {
        return tags;
    }

    public Map<Currency, Quote> getQuotes() {
        return quotes;
    }

    public List<Quote> getQuotesAsList() {
        if (quotes == null) {
            return null;
        }
        return new ArrayList<>(quotes.values());
    }

    public void clearQuote() {
        quotes.clear();
    }

    public boolean hasQuote() {
        if (quotes == null) {
            return false;
        }
        return !quotes.isEmpty();
    }

    public boolean hasQuote(String currency) {
        if (quotes == null) {
            return false;
        }
        return quotes.containsKey(Currency.valueOf(currency));
    }

    public boolean hasQuote(Currency currency) {
        if (quotes == null) {
            return false;
        }
        return quotes.containsKey(currency);
    }

    public boolean hasQuote(Currency[] currencies) {
        if (quotes == null) {
            return false;
        }
        for (Currency currency : currencies) {
            if (!quotes.containsKey(currency)) {
                return false;
            }
        }
        return true;
    }

    public void addQuotes(List<Quote> quotes) {
        for (Quote quote : quotes) {
            addQuote(quote);
        }
    }

    public Quote getQuote(Currency currency) {
        if (quotes != null) {
            return quotes.get(currency);
        }
        return null;
    }

    public Quote getUsdQuote() {
        return getQuote(Currency.USD);
    }

    public Quote getBtcQuote() {
        return getQuote(Currency.BTC);
    }
}

File: lca/src/main/kotlin/com/dreampany/lca/data/model/Coin.kt
package com.dreampany.lca.data.model

import android.os.Parcel
import android.os.Parcelable
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.dreampany.frame.data.model.BaseKt
import com.dreampany.frame.util.TimeUtil
import com.dreampany.lca.data.enums.CoinSource
import com.dreampany.lca.data.enums.Currency
import com.dreampany.lca.misc.Constants
import com.google.common.collect.Maps
import com.google.firebase.database.Exclude
import com.google.firebase.database.IgnoreExtraProperties
import com.google.firebase.database.PropertyName
import java.io.Serializable
import java.util.*

/**
 * Created by roman on 2019-07-28
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@IgnoreExtraProperties
@Entity(
    indices = [Index(
        value = [Constants.Coin.ID],
        unique = true
    )],
    primaryKeys = [Constants.Coin.ID]
)
class Coin(
    override var time: Long,
    override var id: String
) : BaseKt() {

    var source: CoinSource? = null
    var name: String? = null
    var symbol: String? = null
    var slug: String? = null
    var rank: Int = 0
    @PropertyName(Constants.Coin.MARKET_PAIRS)
    private var marketPairs: Int = 0
    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    private var circulatingSupply: Double = 0.0
    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    private var totalSupply: Double = 0.0
    @PropertyName(Constants.Coin.MAX_SUPPLY)
    private var maxSupply: Double = 0.0
    @PropertyName(Constants.Coin.LAST_UPDATED)
    private var lastUpdated: Long = 0
    @PropertyName(Constants.Coin.DATE_ADDED)
    private var dateAdded: Long = 0
    var tags: MutableList<String>? = null
    @Ignore
    @Exclude
    private var quotes: MutableMap<Currency, Quote>? = null

    @Ignore
    constructor() : this("") {

    }

    constructor(id: String) : this(TimeUtil.currentTime(), id) {

    }

    @Ignore
    private constructor(parcel: Parcel) : this(parcel.readLong(), parcel.readString()!!) {
        source = parcel.readParcelable(CoinSource::class.java.classLoader)
        name = parcel.readString()
        symbol = parcel.readString()
        slug = parcel.readString()
        rank = parcel.readInt()
        marketPairs = parcel.readInt()
        circulatingSupply = parcel.readDouble()
        totalSupply = parcel.readDouble()
        maxSupply = parcel.readDouble()
        lastUpdated = parcel.readLong()
        dateAdded = parcel.readLong()
        tags = parcel.createStringArrayList()
        quotes = parcel.readSerializable() as MutableMap<Currency, Quote>?

    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeLong(time)
        dest.writeString(id)
        dest.writeParcelable(source, flags)
        dest.writeString(name)
        dest.writeString(symbol)
        dest.writeString(slug)
        dest.writeInt(rank)
        dest.writeInt(marketPairs)
        dest.writeDouble(circulatingSupply)
        dest.writeDouble(totalSupply)
        dest.writeDouble(maxSupply)
        dest.writeLong(lastUpdated)
        dest.writeLong(dateAdded)
        dest.writeStringList(tags)
        dest.writeSerializable(quotes as Serializable)
    }

    companion object CREATOR : Parcelable.Creator<Coin> {
        override fun createFromParcel(parcel: Parcel): Coin {
            return Coin(parcel)
        }

        override fun newArray(size: Int): Array<Coin?> {
            return arrayOfNulls(size)
        }
    }

    override fun toString(): String {
        return "Coin ($id) == $id"
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    fun setMarketPairs(marketPairs: Int) {
        this.marketPairs = marketPairs
    }

    @PropertyName(Constants.Coin.MARKET_PAIRS)
    fun getMarketPairs(): Int {
        return marketPairs
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    fun setCirculatingSupply(circulatingSupply: Double) {
        this.circulatingSupply = circulatingSupply
    }

    @PropertyName(Constants.Coin.CIRCULATING_SUPPLY)
    fun getCirculatingSupply(): Double {
        return circulatingSupply
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    fun setTotalSupply(totalSupply: Double) {
        this.totalSupply = totalSupply
    }

    @PropertyName(Constants.Coin.TOTAL_SUPPLY)
    fun getTotalSupply(): Double {
        return totalSupply
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    fun setMaxSupply(maxSupply: Double) {
        this.maxSupply = maxSupply
    }

    @PropertyName(Constants.Coin.MAX_SUPPLY)
    fun getMaxSupply(): Double {
        return maxSupply
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    fun setLastUpdated(lastUpdated: Long) {
        this.lastUpdated = lastUpdated
    }

    @PropertyName(Constants.Coin.LAST_UPDATED)
    fun getLastUpdated(): Long {
        return lastUpdated
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    fun setDateAdded(dateAdded: Long) {
        this.dateAdded = dateAdded
    }

    @PropertyName(Constants.Coin.DATE_ADDED)
    fun getDateAdded(): Long {
        return dateAdded
    }

    @Exclude
    fun getLastUpdatedDate(): Date {
        return Date(getLastUpdated())
    }

    fun addQuote(quote: Quote) {
        if (quotes == null) {
            quotes = Maps.newHashMap()
        }
        quotes!![quote.currency] = quote
    }

    @Exclude
    fun getQuotes(): Map<Currency, Quote>? {
        return quotes
    }

    @Exclude
    fun getQuotesAsList(): List<Quote>? {
        return if (quotes == null) {
            null
        } else ArrayList(quotes!!.values)
    }

    fun clearQuote() {
        quotes?.clear()
    }

    fun hasQuote(): Boolean {
        return !quotes.isNullOrEmpty()
    }

    fun hasQuote(currency: String): Boolean {
        return if (quotes == null) {
            false
        } else quotes!!.containsKey(Currency.valueOf(currency))
    }

    fun hasQuote(currency: Currency): Boolean {
        return if (quotes == null) {
            false
        } else quotes!!.containsKey(currency)
    }

    fun hasQuote(currencies: Array<Currency>): Boolean {
        if (quotes == null) {
            return false
        }
        for (currency in currencies) {
            if (!quotes!!.containsKey(currency)) {
                return false
            }
        }
        return true
    }

    fun addQuotes(quotes: List<Quote>) {
        for (quote in quotes) {
            addQuote(quote)
        }
    }

    fun getQuote(currency: Currency): Quote? {
        return if (quotes != null) {
            quotes!!.get(currency)
        } else null
    }

    @Exclude
    fun getLatestQuote(): Quote? {
        var latest: Quote? = null
        quotes?.forEach {
            if (latest == null || latest!!.time < it.value.time) {
                latest = it.value
            }
        }
        return latest
    }
}
